plugins/out_clp_s3/flush/flush.go

```diff
@@ -21,111 +22,226 @@ import (
-		ts, record, err := decoder.GetRecord(dec)
-		if err == io.EOF {
-			// Chunk decoding finished. Break out of loop and send log events to output.
-			break
-		} else if err != nil {
-			err = fmt.Errorf("error decoding data from stream: %w", err)
-			return output.FLB_ERROR, err
-		}
+	tag, ok := ctx.Tags[tagKey]
 
-		timestamp := decodeTs(ts)
-		msg, err := getMessage(record, ctx.Config)
+	// If tag does not exist yet, create new buffers and tag. If UseDiskBuffer is set, buffers are
+	// created on disk and are used to buffer Fluent Bit chunks. If UseDiskBuffer is off, buffer is
+	// in memory and chunks are not buffered.
+	if !ok {
+		irBuf, zstdBuf, err := newBuffers(
+			ctx.Config.UseDiskBuffer,
+			ctx.Config.DiskBufferPath,
+			tagKey,
+		)
 		if err != nil {
-			err = fmt.Errorf("failed to get message from record: %w", err)
-			return output.FLB_ERROR, err
+			return output.FLB_RETRY, fmt.Errorf("error creating buffers: %w", err)
 		}
 
-		event := ffi.LogEvent{
-			LogMessage: msg,
-			Timestamp:  ffi.EpochTimeMs(timestamp.UnixMilli()),
+		tag, err = NewTag(
+			tagKey,
+			ctx.Config.TimeZone,
+			size,
+			ctx.Config.UseDiskBuffer,
+			irBuf,
+			zstdBuf,
+		)
+
+		if err != nil {
+			return output.FLB_RETRY, fmt.Errorf("error creating tag: %w", err)
 		}
-		logEvents = append(logEvents, event)
+		ctx.Tags[tagKey] = tag
 	}
 
-	var buf bytes.Buffer
-
-	zstdWriter, err := zstd.NewWriter(&buf)
+	err = tag.Writer.WriteIrZstd(logEvents)
 	if err != nil {
-		err = fmt.Errorf("error opening zstd writer: %w", err)
-		return output.FLB_RETRY, err
+		return output.FLB_ERROR, err
 	}
 
-	// IR buffer using bytes.Buffer internally, so it will dynamically grow if undersized. Using
-	// FourByteEncoding as default encoding.
-	irWriter, err := ir.NewWriterSize[ir.FourByteEncoding](length, ctx.Config.TimeZone)
+	readyToUpload, err := checkUploadCriteria(
+		tag,
+		ctx.Config.UseDiskBuffer,
+		ctx.Config.UploadSizeMb,
+	)
 	if err != nil {
-		err = fmt.Errorf("error opening IR writer: %w", err)
-		return output.FLB_RETRY, err
+		return output.FLB_ERROR, fmt.Errorf("error checking upload criteria: %w", err)
 	}
 
-	err = writeIr(irWriter, logEvents)
-	if err != nil {
-		err = fmt.Errorf("error while encoding IR: %w", err)
-		return output.FLB_ERROR, err
+	if readyToUpload {
+		err := FlushZstdToS3(tag, ctx)
+		if err != nil {
+			return output.FLB_ERROR, fmt.Errorf("error flushing Zstd buffer to s3: %w", err)
+		}
 	}
 
-	// Write zstd compressed IR to file.
-	_, err = irWriter.CloseTo(zstdWriter)
+	return output.FLB_OK, nil
+}
+
+// Creates a new tag containing a new [irzstd.IrZstdWriter].
+//
+// Parameters:
+//   - tagKey: Fluent Bit tag
+//   - timezone: Time zone of the log source
+//   - size: Byte length
+//   - useDiskBuffer: On/off for disk buffering
+//   - irBuffer: Buffer for IR
+//   - ZstdBuffer: Buffer for Zstd compressed IR
+//
+// Returns:
+//   - tag: Tag resources and metadata
+//   - err: Error creating new writer
+func NewTag(
```

Comment:
```suggestion
func newTag(
```
This doesn't appear used outside `flush` can it be private?